Controller Support

fleisch11 opened the pull request:
made some last minute changes.
it works perfect in chrome and ok in firefox.
i tested it with a switch pro controller. (it should work with other controllers, but i dont have another controller to test it)

elliesec commented:
I can pull the branch down and give it a test with some other controllers (XBox One/PS4) to verify that it's working correctly with them, but I won't be able to until some time next week.

fleisch11 commented:
> I can pull the branch down and give it a test with some other controllers (XBox One/PS4) to verify that it's working correctly with them, but I won't be able to until some time next week.

thanks for doing that in advance and thanks for telling me that.

elliesec reviewed (changes requested):
Thanks for submitting this, and sorry it's taken a while for me to be able to give it a proper test.

I've now pulled it down and tested with an XBox One controller and a PS4 controller. I'm seeing some pretty substantial drift with both - the XBox controller consistently drifts downwards and to the right, whereas the PS4 controller drifts in whichever direction the left stick was moved in last. Aside from the drift, I've not noticed any major functionality issues.

Aside from that though, there are a few things that I think need addressing code-wise before this is ready:
* Placeholder icons in the preference pages (if you're not confident in being able to produce icons, I'd suggest asking if anyone would be interested in making icons for the feature in Discord)
* The user-facing text in the calibration steps on the preference page should be moved to `Text_Preference.csv` to ensure that translators can provide translations.
* In general, prefer the use of `MouseIn(x, y, w, h)` over this:
```js
if (MouseX >= x && MouseX < x + w && MouseY >= y && MouseY < y + h)
```
* `ControllerSupport.js` would benefit from some JSDoc on functions - it will help Ben and the developers to maintain the code in the long run
  * In general there are a lot of single-letter variables in `ControllerSupport.js` - naming them a bit more descriptively, or adding comments to explain what's going on would help devs to understand the code better
* It would generally be preferable to remove commented-out code & `console.log` statements - do this in a separate commit if you feel you might need them, and then you'll still be able to roll back to retrieve them

Thanks again. It would be awesome if we could get this to a point where we can merge it!

fleisch11 commented:
ok, I'll work on it as soon as i can.
I am not sure, wther i can adress the drifting issue, since it is on controlers, that i dont have, but i have ordered an xbox one controller, that'll arrive in about 10 days

fleisch11 commented:
done:
-the drifting issue should be fixed (i have another controller with the same issue)
-added sensitivity setting
-moved preference text to csv
-removed the commented code at the end of controllersupport.js
-added jsdoc to functions in controllersupport.js

to do:
-resolve the drawing conflict
-replace [if (mouse >=x...] with [mousein(...]
-get an icon

elliesec commented:
Oh, one other thing (sorry - forgot to mention this the first time around). Because a lot of screen/file-related variables are global, we conventionally prefix them with the name of the file they're in to prevent naming conflicts - it would be sensible to do the same in `ControllerSupport.js` and with the `CalibrationStage` variable added to `Preference.js`, i.e.:
```js
// ControllerSupport.js
var ControllerSupportButtonsX = [];
var ControllerSupportButtonsY = [];
...

// Preference.js
var PreferenceCalibrationStage = 0;
```
I appreciate it's more verbose, but it reduces the likelihood of files overwriting global variables set by other files, which can sometimes be a tricky problem to identify.

Thanks for the changes so far though - I'll pull it down and give it another test soon.

fleisch11 commented:
ok, it should be finished now.

elliesec commented:
I'm afraid I'm still have drift issues with both controllers - they're behaving pretty much as before, although the PS4 controller seems a little more steady than before - I can now get it to stay still with some effort.

In addition, the two preference menu calibration options don't seem to work any more. Entering the "Map Buttons" screen seems to work, but upon hitting the first button, the screen appears to skip very rapidly through all the subsequent steps and then leaves the preference menu. Selecting the "Map Sticks" button doesn't enter a new subscreen but seems to send the cursor up to the far right of the screen. It then stays there regardless of what I do (moving the mouse momentarily brings it back to the mouse position, but it will launch itself back to the top right corner very quickly once the mouse stops moving).

Nina-1474 commented:
If I'm reading it correctly, the deadzone value is 0.01, which seems too small. My Xbox controller's joystick at rest in the centre gives an average of about 0.06 in both directions, and occasionally exceeds 0.1. This threshold should probably be either configurable or a good bit higher.

fleisch11 commented:
ok, I'll get to work

fleisch11 commented:
about map buttons:
i had the same issue, but i dont anymore, though i dont know how to fix it (if i didnt know any better, i'd think you didnt use the most recent version) (for now i'll add a timer instead)

about the stick configuring: I'll work on that

about deadzone: i'll add that

fleisch11 commented:
should fix the stick configuring

added deadzone

i do not have the button issue, so i dont know how to work on that (if the issue still exists, i'll get to that once my xbox one controller arrives (in about 10-20 days))

fleisch11 commented:
the stick confuguration should be fixed

added deadzone confuguration

i do not have the button issue, so i dont know how to work on that (if the issue still exists, i'll get to that once my xbox one controller arrives (in about 10-20 days))

elliesec commented:
Hmm, I was seeing the button issues reliably on both controllers. I suspect it might be that the first prompt is asking for me to press "A", and the A on an XBox controller (I tried configuring both controllers according to an XBox layout because PS4 doesn't have A/B/Y) corresponds to B on a Switch controller, which might be triggering the exit functionality? That's just a guess though - if that doesn't give you a lead, I'll check it out and have a little debug at some point.

elliesec commented:
@Ben987 - FYI, I think this is very close now. I'm going to pull it down again soon and give it one more test run, and will update you once I have.
@fleisch11 - in the meantime, it's probably worth resolving the merge conflicts on this.

fleisch11 commented:
ok, i am almost done now.
i just need to redo the drawing.js .

fleisch11 commented:
ok, i am finished now.

-i redid the drawing.js, so it works with the current version
-i moved the lines of code that draw the cyan rectangle to the timer.js (that way it always gets drawn)
-i resolved the conflicts

fleisch11 commented:
ok, i am finished now.

-i redid the drawing.js, so it works with the current version
-i moved the lines of code that draw the cyan rectangle to the timer.js (that way it always gets drawn)
-i resolved the conflicts
-i fixed an oversight that caused the dpadright value to not be stored

fleisch11 commented:
resolved the new conflicts

fleisch11 commented:
resolved the new conflicts

elliesec reviewed (approved):
I've given this a pretty thorough testing now PS4, XBox One and Switch Pro controllers, and am happy that it's now in a good state and everything looks to be working as expected.

@Ben987 - if you're happy to have the functionality in the game, it should now be ready for merging.

Ben987 commented:
@fleisch11 , maybe it's a github visual glitch, but it looks like your text indentation is different than what we currently have in the code.  Can you check to make sure you use regular tabs to indent your code?  Thanks a lot.

Ben987 commented:
I will prepare a quick a survey on Discord to see if players would like to have joystick support in the game :)

fleisch11 commented:
if you are talking about the part in dialog.js: it's a github visual glitch
if you are talking about another part: probably too, but i'll check it (and if necessary fix it) as soon as i can (in at most 12 hours)

elliesec commented:
Looks like the lines that have changed are using 4 spaces for indentation rather than tabs. Keeping things at least consistent within the same file would be good.

fleisch11 commented:
it would appear, that visual studio autoaticly converts tab to 4 spaces.
i dont have the time at this moment, but I'll fix it in about 6 hours, if you want me too

fleisch11 commented:
i will be able to fix it in about 3 hours.
its because visual studio automaticly used 4 spaces, even if i press tab.
i changed the settings, but dont have time to fix everything right now

elliesec commented:
I've set my IDE to use tabs for the BC project, which is probably something worth doing. No problem though, there's no rush. Whenever you get the time.

Nina-1474 commented on BondageClub/Scripts/ExtendedItem.js:
```diff
@@ -148,25 +148,28 @@ function ExtendedItemDraw(Options, DialogPrefix, OptionsPerPage, ShowImages = tr
+        if (ShowImages) DrawImage("Screens/Inventory/" + Asset.Group.Name + "/" + Asset.Name + "/" + Option.Name + ".png", X + 2, Y);
+        DrawTextFit(DialogFind(Player, DialogPrefix + Option.Name), X + 112, Y + 30 + ImageHeight, 225, "black");
+        if (ControllerActive == true) {
+            DrawTextFit(X + 112, Y + 30 + ImageHeight);
```
I'm guessing this is supposed to be `setButton` rather than `DrawTextFit`.

fleisch11 replied:
darn, you're right

fleisch11 commented:
ok, all the indents are now tabs instead of 4 spaces
